Fixes and tests for raft-based topology changes

kbr-scylla opened the pull request:
Fix two issues with the replace operation introduced by recent PRs.

Add a test which performs a sequence of basic topology operations (bootstrap,
decommission, removenode, replace) in a new suite that enables the `raft`
experimental feature (so that the new topology change coordinator code is used).

Fixes: #13651

gleb-cloudius commented on test/topology_experimental_raft/test_topology_ops.py:
```diff
@@ -0,0 +1,49 @@
+#
```
I was thinking that we can somehow run the existing topology suit in two modes. Was discussing how to do it best with @kostja. He had some ideas.

kbr-scylla replied:
Is that a blocker for the PR?

BTW. The test I added here would fail in a flaky way if we tried to run it in gossiper mode, with errors such as "decommission is still running" (when you try to perform the next decommission) etc.

kbr-scylla replied:
The ideas are compatible. E.g. have markers like `raft_topology_only` and `gossip_topology_only`. Unmarked tests run in both modes.

gleb-cloudius replied:
They are. But we still managed to have an argument about it.

kostja replied:
I think they aren't, since my suggsetion is that there is no default. One has to specify modes explicitly for each test. This way we don't  bloat the coverage and only run relevant tests in both modes.

kbr-scylla commented:
@tgrabiec please review/merge

kbr-scylla commented:
The test was failing on next promotion: https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/5918/artifact/testlog/x86_64/debug/topology_experimental_raft.test_topology_ops.1.log

I dequeued the PR.
Investigating

kbr-scylla commented:
https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/5918/artifact/testlog/x86_64/debug/scylla-534.log

```
WARN  2023-04-25 09:26:58,944 [shard 0] storage_service - raft topology: cannot map cbad1ca0-beb7-4d60-a328-9b4ad88db710 to ip, retrying.
WARN  2023-04-25 09:26:59,949 [shard 0] storage_service - (rate limiting dropped 175 similar messages) raft topology: cannot map cbad1ca0-beb7-4d60-a328-9b4ad88db710 to ip, retrying.
WARN  2023-04-25 09:27:00,951 [shard 0] storage_service - (rate limiting dropped 183 similar messages) raft topology: cannot map cbad1ca0-beb7-4d60-a328-9b4ad88db710 to ip, retrying.
WARN  2023-04-25 09:27:01,954 [shard 0] storage_service - (rate limiting dropped 184 similar messages) raft topology: cannot map cbad1ca0-beb7-4d60-a328-9b4ad88db710 to ip, retrying.
...
```

kbr-scylla commented:
Deducing from the logs:
```
INFO  2023-04-25 09:26:54,202 [shard 0] raft_group0 - setup_group0: ensuring that the cluster has fully upgraded to use Raft...
WARN  2023-04-25 09:26:54,816 [shard 0] gossip - Fail to send EchoMessage to 127.142.143.54: seastar::rpc::closed_error (connection is closed)
INFO  2023-04-25 09:26:54,824 [shard 0] gossip - InetAddress 127.142.143.51 is now UP, status = NORMAL
INFO  2023-04-25 09:26:54,825 [shard 0] gossip - InetAddress 127.142.143.37 is now UP, status = NORMAL
INFO  2023-04-25 09:26:54,839 [shard 0] gossip - InetAddress 127.142.143.54 is now DOWN, status = shutdown
INFO  2023-04-25 09:26:55,059 [shard 0] raft_group_registry - marking Raft server da1d1ebf-3bef-47af-b72a-172a8c1068f5 as alive for raft groups
INFO  2023-04-25 09:26:55,066 [shard 0] raft_group_registry - marking Raft server 8058ce55-6363-4a8e-9bab-23d5085478fb as alive for raft groups
INFO  2023-04-25 09:26:55,070 [shard 0] raft_group_registry - marking Raft server 8ed2fe09-287e-46db-b303-2eb5bcc570e9 as alive for raft groups
INFO  2023-04-25 09:26:55,072 [shard 0] schema_tables - Creating keyspace system_distributed
INFO  2023-04-25 09:26:55,582 [shard 0] schema_tables - Schema version changed to 82b0cad2-7d17-3f21-b083-5acebf38f64d
INFO  2023-04-25 09:26:55,616 [shard 0] schema_tables - Creating keyspace system_distributed_everywhere
INFO  2023-04-25 09:26:56,182 [shard 0] schema_tables - Schema version changed to 4217928c-ec3e-3ce2-8671-106268ae6359
INFO  2023-04-25 09:26:56,846 [shard 0] schema_tables - Creating system_distributed.cdc_streams_descriptions_v2 id=0bf73fd7-65b2-36b0-85e5-658131d5df36 version=f9f0c8dc-afda-31b6-9a62-fa952635e087
INFO  2023-04-25 09:26:56,858 [shard 0] schema_tables - Creating system_distributed_everywhere.cdc_generation_descriptions_v2 id=234d2227-dd63-3d37-ac5f-c013e2ea9e6e version=fb878fc3-505f-3ee9-b701-dd61b3d40a5a
INFO  2023-04-25 09:26:56,870 [shard 0] schema_tables - Creating system_distributed.view_build_status id=5582b59f-8e4e-35e1-b913-3acada51eb04 version=03941b0a-6723-3482-bae5-2f8b2944eafe
INFO  2023-04-25 09:26:56,881 [shard 0] schema_tables - Creating system_distributed.service_levels id=b8c556bd-212d-37ad-9484-690c73a5994b version=c06a55bc-9aec-36ff-91d8-761c5d08aa9f
INFO  2023-04-25 09:26:56,889 [shard 0] schema_tables - Creating system_distributed.cdc_generation_timestamps id=fdf455c4-cfec-3e00-9719-d7a45436c89d version=30b59430-8d40-3968-88f5-664df646d4a5
INFO  2023-04-25 09:26:58,062 [shard 0] schema_tables - Schema version changed to 491eac62-9cdb-363f-8f91-7f99bdddb1ef
INFO  2023-04-25 09:26:58,190 [shard 0] schema_tables - Altering system_distributed.service_levels id=b8c556bd-212d-37ad-9484-690c73a5994b version=fcbf5db3-41fb-37a0-a39c-58effa7a77a7
INFO  2023-04-25 09:26:58,202 [shard 0] query_processor - Column definitions for system_distributed.service_levels changed, invalidating related prepared statements
INFO  2023-04-25 09:26:58,207 [shard 1] query_processor - Column definitions for system_distributed.service_levels changed, invalidating related prepared statements
INFO  2023-04-25 09:26:58,894 [shard 0] schema_tables - Schema version changed to 357632c4-7988-3bbb-8dc2-540f038b02e2
WARN  2023-04-25 09:26:58,944 [shard 0] storage_service - raft topology: cannot map cbad1ca0-beb7-4d60-a328-9b4ad88db710 to ip, retrying.
```

the replacing server (scylla-534) hanged at the initial `read_barrier` that happens after we join group 0 (`raft_group0::setup_group0`):
```cpp
    group0_log.info("setup_group0: ensuring that the cluster has fully upgraded to use Raft...");
    auto& group0_server = _raft_gr.group0();
    group0_members members0{group0_server, _raft_gr.address_map()};

    // Perform a Raft read barrier so we know the set of group 0 members and our group 0 state is up-to-date.
    co_await group0_server.read_barrier(&_abort_source);
```

I think I know what could be happening...

There was no log truncation, so the joining server obtains the entire Raft log and tries to apply its entries in order. This includes early entries that describe a topology state which contains some nodes which have been removed by now. Unfortunately, applying group 0 entries requires reloading the topology state (`storage_service::topology_state_load`), which requires mapping the Host/Raft IDs of nodes present in this topology to their IPs (`id2ip` lambda) because `token_metadata` uses IPs internally. But the cluster is no longer gossiping these IPs, because these nodes are no longer members in the newest topology!

This also explains why the joining node considers some nodes that are actually present to be `FatClient`s - according to this node, which has only applied old group 0 entries, these nodes (which are NORMAL in the most recent topology) are not members of the topology.

cc @gleb-cloudius @kostja @tgrabiec 

At this moment I see no other way to fix this than solving https://github.com/scylladb/scylladb/issues/6403. We must not try to map IDs to IPs when applying group 0 entries, because some IDs may be missing their mappings when we apply old entries.

kbr-scylla commented:
For now I modified the test to not stumble on this problem. The idea is that we don't try to join new nodes to the cluster after we removed some nodes from the cluster - the existing nodes don't have to apply those old topology entries, they are using the new topology.

The test provides less coverage but it's better than no test.
The old test is on this branch: https://github.com/kbr-scylla/scylladb/tree/test-experimental-raft-v1

kbr-scylla commented:
But there's no guarantee how Raft will group entries into batches and provide them into `state_machine::apply`.
Also even if we got all entries in a single batch, is it possible to correctly squash them when dealing with different entry types? For example if there are schema entries interleaved with topology entries. What if applying a topology entry depends on an earlier schema entry etc.?

gleb-cloudius commented:
> But there's no guarantee how Raft will group entries into batches and provide them into `state_machine::apply`.

That is why I did not say this is a proper solution.

> Also even if we got all entries in a single batch, is it possible to correctly squash them when dealing with different entry types? For example if there are schema entries interleaved with topology entries. What if applying a topology entry depends on an earlier schema entry etc.?

I hope we will not have such circular dependencies. It make sense for schema mutation to depend on the topology, so if we apply topology first and then schema it should work. If we introduce topology on schema dependency then there will be a problem indeed.

kbr-scylla commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/883/

https://github.com/scylladb/scylladb/issues/13211

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/889/

kbr-scylla commented:
@tgrabiec please review/merge again
